experimental/dds/tree2/src/test/shared-tree-core/editManager.spec.ts

```diff
@@ -404,57 +387,101 @@ describe("EditManager", () => {
+					{ seq: 5, type: "Pull", ref: 4, from: peer2 },
+					{ seq: 6, type: "Pull", ref: 5, from: peer1 },
+					{ seq: 7, type: "Pull", ref: 5, from: peer1 },
+				],
+				new UnrebasableTestChangeRebaser(),
+			);
+			runUnitTestScenario(
+				"Sequenced local changes should not be rebased over prior local changes if those earlier changes were not rebased",
+				[
+					{ seq: 1, type: "Push" },
+					{ seq: 2, type: "Push" },
+					{ seq: 4, type: "Push" },
+					{ seq: 1, type: "Ack" },
+					{ seq: 2, type: "Ack" },
+					{ seq: 3, type: "Pull", ref: 2, from: peer2 },
+					{ seq: 4, type: "Ack" },
+				],
+				new ConstrainedTestChangeRebaser(
+					(change: TestChange, over: TaggedChange<TestChange>): boolean => {
+						// This is the only rebase that should happen
+						assert.deepEqual(change.intentions, [4]);
+						assert.deepEqual(over.change.intentions, [3]);
+						return true;
+					},
+				),
+			);
+			runUnitTestScenario(
+				"Sequenced peer changes should not be rebased over changes from the same peer if those earlier changes were not rebased",
+				[
+					{ seq: 1, type: "Pull", ref: 0, from: peer1 },
+					{ seq: 2, type: "Pull", ref: 0, from: peer1 },
+					{ seq: 3, type: "Pull", ref: 2, from: peer2 },
+					{ seq: 4, type: "Pull", ref: 0, from: peer1 },
+				],
+				new ConstrainedTestChangeRebaser(
+					(change: TestChange, over: TaggedChange<TestChange>): boolean => {
+						// This is the only rebase that should happen
+						assert.deepEqual(change.intentions, [4]);
+						assert.deepEqual(over.change.intentions, [3]);
+						return true;
+					},
+				),
+			);
+		});
+
+		interface Scenario {
+			readonly nbRebased: number;
+			readonly nbTrunk: number;
+		}
+
+		const scenarios: Scenario[] = [
+			{ nbRebased: 1, nbTrunk: 1 },
+			{ nbRebased: 10, nbTrunk: 1 },
+			{ nbRebased: 1, nbTrunk: 10 },
+			{ nbRebased: 7, nbTrunk: 3 },
+		];
+
+		describe("Local commit rebasing", () => {
+			for (const { nbRebased, nbTrunk } of scenarios) {
+				it(`Rebase ${nbRebased} local commits over ${nbTrunk} trunk commits`, () => {
+					const rebaser = new NoOpChangeRebaser();
+					rebaseLocalEditsOverTrunkEdits(nbRebased, nbTrunk, rebaser);
+					assert.equal(rebaser.nbRebased, nbTrunk * nbRebased ** 2);
+					assert.equal(rebaser.nbInverted, nbTrunk * nbRebased);
+					assert.equal(rebaser.nbComposed, nbTrunk * (nbRebased * 2 + 1));
+					assert.equal(rebaser.nbRebaseAnchorCalls, nbTrunk + nbRebased);
+				});
+			}
+		});
+		describe("Peer commit rebasing", () => {
+			for (const { nbRebased, nbTrunk } of scenarios) {
+				it(`Rebase ${nbRebased} peer commits over ${nbTrunk} trunk commits`, () => {
+					const rebaser = new NoOpChangeRebaser();
+					rebasePeerEditsOverTrunkEdits(nbRebased, nbTrunk, rebaser);
+					assert.equal(
+						rebaser.nbRebased,
+						nbTrunk * nbRebased + nbRebased * (nbRebased - 1),
+					);
+					// TODO: Task4664 Prevent quadratic number of inversions by caching inverses
```

Comment:
```suggestion
					// TODO: Prevent quadratic number of inversions by caching inverses
```
Technically we're not supposed to put internal ADO items in public github or source :/. I often find myself wanting to do this, though. I wonder if we can revisit this policy.

Reply:
Nevermind :)

Reply:
Do you have a pointer to where that policy is documented? The rest of the codebase has generally not been always following it--I know I've introduced a few cases. My impression was our policy was that open-source contributors should be able to understand what the "to-do" work was without having access to the backlog item, but convenience links are fine. Ex: things like `TODO:AB#foo: enable this test` or `// this code is for runtime compat and can be removed after LTS exceeds 2.0.0 (tracked by AB#bar)` are OK. Things like `// TODO:AB#500` or `// fix this (AB#3)` are not.

Reply:
Sorry - we talked about this at standup this morning, and it is okay to do (hence my "nevermind"). I was mistaken.